File: superowner_course_compiler.py
import os
import sys
import re
import pypdf

# Define paths relative to the active workspace
workspace_dir = r"C:\Users\priya\.gemini\antigravity\scratch\html-tutorial-codewith-ai"
app_js_path = os.path.join(workspace_dir, "app.js")
index_html_path = os.path.join(workspace_dir, "index.html")

def compile_new_pdf_course(pdf_path, course_title, course_id):
    print(f"\\n🚀 Starting compilation for: {course_title}...")
    
    if not os.path.exists(pdf_path):
        print(f"❌ Error: PDF file not found at path: {pdf_path}")
        return
        
    # 1. Read & Extract PDF
    try:
        reader = pypdf.PdfReader(pdf_path)
        total_pages = len(reader.pages)
        print(f"📖 Extracted {total_pages} pages from PDF.")
    except Exception as e:
        print(f"❌ Error reading PDF: {e}")
        return

    # 2. Read current app.js to get total slide length
    with open(app_js_path, "r", encoding="utf-8") as f:
        app_js_content = f.read()

    # Search for slides count
    existing_slides_count = app_js_content.count("level:")
    print(f"📊 Current slides in database: {existing_slides_count}")
    
    new_slides_js_list = []
    sidebar_links = []
    
    # 3. Parse pages and create slides data
    for i in range(total_pages):
        page = reader.pages[i]
        page_text = page.extract_text() or ""
        lines = [l.strip() for l in page_text.split("\\n") if l.strip()]
        
        slide_index = existing_slides_count + i
        slide_number = slide_index + 1
        
        # Deduce title from first line of text
        title = ""
        header_idx = 0
        if len(lines) > 0:
            for idx, line in enumerate(lines[:3]):
                cleaned = re.sub(r"^[-\\*\\•\\s]+", "", line).strip()
                if cleaned and not cleaned.isdigit() and len(cleaned) > 2 and "VIRTUAL" not in cleaned.upper():
                    title = cleaned
                    header_idx = idx
                    break
        if not title:
            title = f"Topic Overview - Page {i+1}"
            
        if len(title) > 35:
            title = title[:32] + "..."
            
        # Rebrand branding strings
        title = re.sub(r"VIRTUAL CODE", "codewith_ai", title, flags=re.IGNORECASE)
        
        # Create bullet points HTML
        items_html = ""
        items = []
        for line in lines[header_idx:]:
            cleaned_l = re.sub(r"^[-\\*\\•\\s]+", "", line).strip()
            cleaned_l = re.sub(r"VIRTUAL CODE", "codewith_ai", cleaned_l, flags=re.IGNORECASE)
            cleaned_l = cleaned_l.replace("`", "'")
            if cleaned_l:
                items.append(f"<li>{cleaned_l}</li>")
                
        if items:
            items_html = f"<ul>{''.join(items)}</ul>"
        else:
            items_html = f"<p>Detail slides content from course guide.</p>"
            
        # Level grouping calculation
        level_num = min(5, i // max(1, (total_pages // 5)))
        level_name = f"Level {level_num}"
        
        slide_level = f"{course_title} {level_name}"
        
        english_html = f"""
            <div class='english-explanation'>
                <div class='english-header'>🇬🇧 English Explanation</div>
                <p>Welcome to <b>{course_title}</b>. This slide outlines key concepts on <b>{title}</b>.</p>
            </div>
        """
        
        hinglish_html = f"""
            <div class='hinglish-explanation'>
                <div class='hinglish-header'>🇮🇳 Hinglish (आसान भाषा में)</div>
                {items_html}
            </div>
        """
        
        slide_body_content = f"{english_html}\\n            {hinglish_html}"
        
        # Sandbox code generation mockup
        sandbox_code = ""
        if "code" in page_text.lower() or "function" in page_text.lower() or "npm" in page_text.lower():
            sandbox_code = "// Practice Sandbox Code\\\\nconsole.log(\\\"Learning " + title + "!\\\");"
            
        escaped_content = slide_body_content.replace("`", "\\\\`").replace("${", "\\\\${")
        
        new_slides_js_list.append(f"""
    {{
        level: "{slide_level}",
        title: "{title}",
        content: `{escaped_content}`,
        sandboxCode: "{sandbox_code}"
    }}""")
        
        sidebar_links.append(f'<li class="slide-link" data-slide="{slide_index}">{slide_number}. {title}</li>')

    # 4. Generate HTML accordion markup
    accordion_id = f"accordion-{course_id}"
    toggle_idx = app_js_content.count("toggleAccordion(") # Deduce new toggle index
    
    links_html = "\\n                                ".join(sidebar_links)
    new_accordion_html = f"""
                <!-- {course_title} Module -->
                <div class="accordion-item" id="{accordion_id}">
                    <button class="accordion-header" onclick="toggleAccordion({toggle_idx})">
                        <span class="header-title">{course_title}</span>
                        <svg class="chevron" width="16" height="16" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2"><polyline points="6 9 12 15 18 9"/></svg>
                    </button>
                    <div class="accordion-content">
                        <div class="level-group" id="nav-{course_id}-group">
                            <div class="level-title">Complete Course Material</div>
                            <ul class="slide-list">
                                {links_html}
                            </ul>
                        </div>
                    </div>
                </div>
            </nav>
        </aside>
    """

    # 5. Insert accordion markup in index.html (before closing tag of nav)
    with open(index_html_path, "r", encoding="utf-8") as f:
        html_content = f.read()
        
    nav_end_marker = "</nav>\\n        </aside>"
    
    if nav_end_marker in html_content:
        html_content = html_content.replace(nav_end_marker, new_accordion_html)
    else:
        nav_end_marker_alt = "</nav>\\r\\n        </aside>"
        if nav_end_marker_alt in html_content:
            html_content = html_content.replace(nav_end_marker_alt, new_accordion_html)
        else:
            print("⚠️ Warning: Navigation tags could not be traced inside index.html. Manual adjustment required.")
            
    with open(index_html_path, "w", encoding="utf-8") as f:
        f.write(html_content)
    print("✓ Successfully updated index.html sidebar accordion list!")

    # 6. Rebuild app.js slides array and showSlide ranges
    slides_data_end = app_js_content.find("];\\n\\n/* Inject style tag dynamically for checkmarks styling */")
    if slides_data_end == -1:
        slides_data_end = app_js_content.find("];\\r\\n\\r\\n/* Inject style tag dynamically for checkmarks styling */")
    if slides_data_end == -1:
        slides_data_end = app_js_content.find("];\\n\\n/* Application States */")
        
    if slides_data_end != -1:
        new_slides_js = ",\n" + ",\n".join(new_slides_js_list)
        rebuild_app_js = app_js_content[:slides_data_end] + new_slides_js + app_js_content[slides_data_end:]
        
        # Update accordion routing indexes ranges inside showSlide(index)
        accordion_range_search = f"else if (index >= 280) {{\\n        activeAccordionIndex = 7;\\n    }}"
        accordion_range_replace = f"else if (index >= 280 && index < {slide_index + 1}) {{\\n        activeAccordionIndex = 7;\\n    }} else if (index >= {slide_index + 1}) {{\\n        activeAccordionIndex = {toggle_idx};\\n    }}"
        
        rebuild_app_js = rebuild_app_js.replace(accordion_range_search, accordion_range_replace)
        rebuild_app_js = rebuild_app_js.replace(accordion_range_search.replace("\\n", "\\r\\n"), accordion_range_replace.replace("\\n", "\\r\\n"))
        
        with open(app_js_path, "w", encoding="utf-8") as f:
            f.write(rebuild_app_js)
        print("✓ Successfully updated app.js slides database!")
    else:
        print("❌ Error: Could not locate end of slidesData array inside app.js.")
        return

    # 7. Auto bump index.html version cache-buster to reload browser assets
    with open(index_html_path, "r", encoding="utf-8") as f:
        final_html = f.read()
    
    v_match = re.search(r"v=(\d+)", final_html)
    if v_match:
        old_v = int(v_match.group(1))
        new_v = old_v + 1
        final_html = final_html.replace(f"v={old_v}", f"v={new_v}")
        with open(index_html_path, "w", encoding="utf-8") as f:
            f.write(final_html)
        print(f"⚡ Bumped browser cache version buster to v={new_v}!")
        
    print(f"\\n🎉 SUCCESS! {course_title} has been compiled into the Course Hub database!")

if __name__ == "__main__":
    print("👑 codewith_ai Course Hub - SuperOwner CLI compiler utility.")
    pdf = input("Enter absolute PDF file path: ").strip()
    title = input("Enter Course Title (e.g., Python Course): ").strip()
    id_tag = input("Enter unique lowercase Course ID (e.g., python-dev): ").strip()
    
    # Remove quotes from paths
    pdf = pdf.strip('"').strip("'")
    
    compile_new_pdf_course(pdf, title, id_tag)

File: superowner_course_compiler.js
const fs = require('fs');
const path = require('path');
const readline = require('readline');
const pdfParse = require('pdf-parse');

const appJsPath = path.join(__dirname, 'app.js');
const indexHtmlPath = path.join(__dirname, 'index.html');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function askQuestion(query) {
    return new Promise(resolve => rl.question(query, resolve));
}

async function compileNewPdfCourse(pdfPath, courseTitle, courseId) {
    console.log(`\n🚀 Starting compilation for: ${courseTitle}...`);

    if (!fs.existsSync(pdfPath)) {
        console.error(`❌ Error: PDF file not found at path: ${pdfPath}`);
        return;
    }

    try {
        const dataBuffer = fs.readFileSync(pdfPath);
        
        // Custom page text extractor to preserve page boundaries
        const pagesText = [];
        const options = {
            pagerender: function(pageData) {
                return pageData.getTextContent().then(function(textContent) {
                    let text = '';
                    for (let item of textContent.items) {
                        text += item.str + '\n';
                    }
                    pagesText.push(text);
                    return text;
                });
            }
        };

        await pdfParse(dataBuffer, options);
        const totalPages = pagesText.length;
        console.log(`📖 Extracted ${totalPages} pages from PDF.`);
        
        // 2. Read current app.js to get total slide length
        let appJsContent = fs.readFileSync(appJsPath, 'utf8');
        
        // Search for slides count
        const existingSlidesCount = (appJsContent.match(/level:/g) || []).length;
        console.log(`📊 Current slides in database: ${existingSlidesCount}`);

        const newSlidesJsList = [];
        const sidebarLinks = [];

        // 3. Parse pages and create slides data
        for (let i = 0; i < totalPages; i++) {
            const pageText = pagesText[i] || "";
            const lines = pageText.split('\n').map(l => l.trim()).filter(l => l.length > 0);
            
            const slideIndex = existingSlidesCount + i;
            const slideNumber = slideIndex + 1;

            let title = "";
            let headerIdx = 0;
            if (lines.length > 0) {
                for (let idx = 0; idx < Math.min(3, lines.length); idx++) {
                    const cleaned = lines[idx].replace(/^[-\*\•\s]+/, '').trim();
                    if (cleaned && !/^\d+$/.test(cleaned) && cleaned.length > 2 && !cleaned.toUpperCase().includes('VIRTUAL')) {
                        title = cleaned;
                        headerIdx = idx;
                        break;
                    }
                }
            }
            if (!title) {
                title = `Topic Overview - Page ${i + 1}`;
            }
            if (title.length > 35) {
                title = title.substring(0, 32) + "...";
            }
            title = title.replace(/VIRTUAL CODE/gi, 'codewith_ai');

            // Create bullet points HTML
            let items = [];
            const remainingLines = lines.slice(headerIdx);
            for (let line of remainingLines) {
                let cleanedL = line.replace(/^[-\*\•\s]+/, '').trim();
                cleanedL = cleanedL.replace(/VIRTUAL CODE/gi, 'codewith_ai');
                cleanedL = cleanedL.replace(/`/g, "'");
                if (cleanedL) {
                    items.push(`<li>${cleanedL}</li>`);
                }
            }
            let itemsHtml = items.length > 0 ? `<ul>${items.join('')}</ul>` : `<p>Detail slides content from course guide.</p>`;

            const levelNum = Math.min(5, Math.floor(i / Math.max(1, totalPages / 5)));
            const levelName = `Level ${levelNum}`;
            const slideLevel = `${courseTitle} ${levelName}`;

            const englishHtml = `
            <div class='english-explanation'>
                <div class='english-header'>🇬🇧 English Explanation</div>
                <p>Welcome to <b>${courseTitle}</b>. This slide outlines key concepts on <b>${title}</b>.</p>
            </div>
            `;

            const hinglishHtml = `
            <div class='hinglish-explanation'>
                <div class='hinglish-header'>🇮🇳 Hinglish (आसान भाषा में)</div>
                ${itemsHtml}
            </div>
            `;

            const slideBodyContent = `${englishHtml}\n            ${hinglishHtml}`;

            let sandboxCode = "";
            const pageLower = pageText.toLowerCase();
            if (pageLower.includes('code') || pageLower.includes('function') || pageLower.includes('npm')) {
                sandboxCode = `// Practice Sandbox Code\\nconsole.log(\\"Learning ${title}!\\");`;
            }

            const escapedContent = slideBodyContent.replace(/`/g, '\\`').replace(/\$\{/g, '\\${');

            newSlidesJsList.push(`
    {
        level: "${slideLevel}",
        title: "${title}",
        content: \`${escapedContent}\`,
        sandboxCode: "${sandboxCode}"
    }`);

            sidebarLinks.push(`<li class="slide-link" data-slide="${slideIndex}">${slideNumber}. ${title}</li>`);
        }

        // 4. Generate HTML accordion markup
        const accordionId = `accordion-${courseId}`;
        const toggleIdx = (appJsContent.match(/toggleAccordion\(/g) || []).length;

        const linksHtml = sidebarLinks.join('\n                                ');
        const newAccordionHtml = `
                <!-- ${courseTitle} Module -->
                <div class="accordion-item" id="${accordionId}">
                    <button class="accordion-header" onclick="toggleAccordion(${toggleIdx})">
                        <span class="header-title">{courseTitle}</span>
                        <svg class="chevron" width="16" height="16" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2"><polyline points="6 9 12 15 18 9"/></svg>
                    </button>
                    <div class="accordion-content">
                        <div class="level-group" id="nav-${courseId}-group">
                            <div class="level-title">Complete Course Material</div>
                            <ul class="slide-list">
                                ${linksHtml}
                            </ul>
                        </div>
                    </div>
                </div>
            </nav>
        </aside>
        `;

        // 5. Insert accordion markup in index.html
        let htmlContent = fs.readFileSync(indexHtmlPath, 'utf8');
        const navEndMarker = '</nav>\n        </aside>';
        const navEndMarkerAlt = '</nav>\r\n        </aside>';

        if (htmlContent.includes(navEndMarker)) {
            htmlContent = htmlContent.replace(navEndMarker, newAccordionHtml);
        } else if (htmlContent.includes(navEndMarkerAlt)) {
            htmlContent = htmlContent.replace(navEndMarkerAlt, newAccordionHtml);
        } else {
            console.warn("⚠️ Warning: Navigation tags could not be traced inside index.html. Manual adjustment required.");
        }

        fs.writeFileSync(indexHtmlPath, htmlContent, 'utf8');
        console.log("✓ Successfully updated index.html sidebar accordion list!");

        // 6. Rebuild app.js slides array and showSlide ranges
        let slidesDataEnd = appJsContent.indexOf('];\n\n/* Inject style tag dynamically for checkmarks & theme custom styling */');
        if (slidesDataEnd === -1) {
            slidesDataEnd = appJsContent.indexOf('];\r\n\r\n/* Inject style tag dynamically for checkmarks & theme custom styling */');
        }
        if (slidesDataEnd === -1) {
            slidesDataEnd = appJsContent.indexOf('];\n\n/* Application States */');
        }

        if (slidesDataEnd !== -1) {
            const newSlidesJs = ",\n" + newSlidesJsList.join(',\n');
            let rebuildAppJs = appJsContent.substring(0, slidesDataEnd) + newSlidesJs + appJsContent.substring(slidesDataEnd);

            const lastSlideIndex = existingSlidesCount + totalPages - 1;
            const accordionRangeSearch = `else if (index >= 280) {\n        activeAccordionIndex = 7;\n    }`;
            const accordionRangeSearchAlt = `else if (index >= 280) {\r\n        activeAccordionIndex = 7;\r\n    }`;
            const accordionRangeReplace = `else if (index >= 280 && index < ${lastSlideIndex + 1}) {\n        activeAccordionIndex = 7;\n    } else if (index >= ${lastSlideIndex + 1}) {\n        activeAccordionIndex = ${toggleIdx};\n    }`;

            if (rebuildAppJs.includes(accordionRangeSearch)) {
                rebuildAppJs = rebuildAppJs.replace(accordionRangeSearch, accordionRangeReplace);
            } else if (rebuildAppJs.includes(accordionRangeSearchAlt)) {
                rebuildAppJs = rebuildAppJs.replace(accordionRangeSearchAlt, accordionRangeReplace.replace(/\n/g, '\r\n'));
            }

            fs.writeFileSync(appJsPath, rebuildAppJs, 'utf8');
            console.log("✓ Successfully updated app.js slides database!");
        } else {
            console.error("❌ Error: Could not locate end of slidesData array inside app.js.");
            return;
        }

        // 7. Auto bump index.html version cache-buster
        let finalHtml = fs.readFileSync(indexHtmlPath, 'utf8');
        const vMatch = finalHtml.match(/v=(\d+)/);
        if (vMatch) {
            const oldV = parseInt(vMatch[1]);
            const newV = oldV + 1;
            finalHtml = finalHtml.replace(new RegExp(`v=${oldV}`, 'g'), `v=${newV}`);
            fs.writeFileSync(indexHtmlPath, finalHtml, 'utf8');
            console.log(`⚡ Bumped browser cache version buster to v=${newV}!`);
        }

        console.log(`\n🎉 SUCCESS! ${courseTitle} has been compiled into the Course Hub database!`);

    } catch (err) {
        console.error(`❌ Error compiling PDF: ${err.message}`);
    }
}

async function main() {
    console.log("👑 codewith_ai Course Hub - SuperOwner JS CLI compiler utility.");
    const pdfPath = (await askQuestion("Enter absolute PDF file path: ")).trim().replace(/['"]/g, '');
    const courseTitle = (await askQuestion("Enter Course Title (e.g., Python Course): ")).trim();
    const courseId = (await askQuestion("Enter unique lowercase Course ID (e.g., python-dev): ")).trim();

    rl.close();
    await compileNewPdfCourse(pdfPath, courseTitle, courseId);
}

main();
